File: dataset/features.py
"""The plan's 29 engineered features: links, money, pressure, credentials, brands,
opt-out language, text shape and relational openers.

Every feature comes from the message body alone, so it works on the UCI and NUS rows
that ship no sender field. sender_type is deliberately excluded: only 955 rows have
one and they all come from SmishTank, so a model given it would learn provenance
rather than smishing. It belongs in the app as a rule.

The word lists live in features.yaml; the logic is here. No pandas above `build()`,
so merge_datasets.py can import this without it.
"""
import os
import re

import yaml

HERE = os.path.dirname(os.path.abspath(__file__))

with open(os.path.join(HERE, "features.yaml"), encoding="utf-8") as fh:
    _CFG = yaml.safe_load(fh)

# ---------------------------------------------------------------- links
# Shared with merge_datasets.py, which uses it for the has_url column.
URL_RE = re.compile(
    r"(https?://|www\.|\b[a-z0-9-]+\.(com|net|org|info|xyz|ru|cn|co|io|me|us|biz|top|"
    r"link|click|vip|shop|site|online|icu|cc|tk|ml|ga|gq|buzz)\b)", re.I)

# Full URL match, used where the domain itself has to be inspected. Group 1 is the
# scheme/www prefix (may be absent), group 2 the host.
URL_FULL_RE = re.compile(
    r"(https?://|www\.)?([a-z0-9][a-z0-9.-]*\.[a-z]{2,})(/\S*)?", re.I)

IP_DOMAIN_RE = re.compile(r"\b(?:\d{1,3}\.){3}\d{1,3}\b")
LEET_RE = re.compile(r"\b[a-z]+[0-9]+[a-z]+\b", re.I)   # M3dicare, Amaz0n
WORD_RE = re.compile(r"\b\w+\b")

# ---------------------------------------------------------------- vocabulary
KNOWN_TLDS = set(_CFG["known_tlds"])
SUSPICIOUS_TLDS = set(_CFG["suspicious_tlds"])
SHORTENERS = set(_CFG["shorteners"])
MISMATCH_BRANDS = _CFG["mismatch_brands"]
# Each family becomes two features: <family>_n (count) and <family>_hit (0/1).
FAMILIES = _CFG["families"]


def _domains(text):
    out = []
    for m in URL_FULL_RE.finditer(text):
        prefix, host = m.group(1), m.group(2).lower().rstrip(".")
        if "." not in host or host.replace(".", "").isdigit():
            continue                                   # "3.50" is not a host
        if not prefix and host.rsplit(".", 1)[-1] not in KNOWN_TLDS:
            continue                                   # "tomorrow.We" is not a host
        out.append(host)
    return out


def _registrable(host):
    parts = host.split(".")
    return ".".join(parts[-2:]) if len(parts) >= 2 else host


def extract_features(text):
    """Return the engineered feature dict for one message body."""
    t = text or ""
    low = t.lower()
    flat = re.sub(r"[^a-z0-9]", "", low)
    n = len(t) or 1
    words = WORD_RE.findall(t)
    domains = _domains(t)
    tlds = {d.rsplit(".", 1)[-1] for d in domains}

    f = {}

    # --- links
    f["has_url"] = int(bool(URL_RE.search(t)))
    f["n_urls"] = len(domains)
    f["has_shortener"] = int(any(_registrable(d) in SHORTENERS for d in domains))
    f["suspicious_tld"] = int(bool(tlds & SUSPICIOUS_TLDS))
    f["ip_as_domain"] = int(bool(IP_DOMAIN_RE.search(t)) and bool(domains))
    f["url_digit_ratio"] = (
        sum(c.isdigit() for d in domains for c in d) / sum(len(d) for d in domains)
        if domains else 0.0)
    f["domain_len_max"] = max((len(d) for d in domains), default=0)
    # brand named in the body but absent from every link in it
    named = [b for b in MISMATCH_BRANDS if b in flat]
    f["brand_domain_mismatch"] = int(
        bool(named) and bool(domains)
        and not any(b in d.replace(".", "").replace("-", "") for b in named for d in domains))

    # --- keyword families
    for fam, terms in FAMILIES.items():
        hits = sum(low.count(term) for term in terms)
        f[f"{fam}_n"] = hits
        f[f"{fam}_hit"] = int(hits > 0)

    # --- text shape
    letters = sum(c.isalpha() for c in t)
    f["length"] = len(t)
    f["n_words"] = len(words)
    f["capital_ratio"] = sum(c.isupper() for c in t) / (letters or 1)
    f["digit_ratio"] = sum(c.isdigit() for c in t) / n
    f["punct_ratio"] = sum(not c.isalnum() and not c.isspace() for c in t) / n
    f["n_exclaim"] = t.count("!")
    f["n_allcaps_words"] = sum(1 for w in words if len(w) > 2 and w.isupper())
    f["has_leetspeak"] = int(bool(LEET_RE.search(t)))
    f["avg_word_len"] = (sum(len(w) for w in words) / len(words)) if words else 0.0

    return f


# Order is fixed by a call on a probe string so downstream arrays stay aligned.
FEATURE_NAMES = list(extract_features("probe http://a.com").keys())

# Readable labels for SHAP plots (plan Phase 6).
FEATURE_LABELS = {
    "has_url": "Contains a link",
    "n_urls": "Number of links",
    "has_shortener": "Link uses a URL shortener",
    "suspicious_tld": "Suspicious top-level domain",
    "ip_as_domain": "IP address used as a domain",
    "url_digit_ratio": "Digits in domain (ratio)",
    "domain_len_max": "Longest domain length",
    "brand_domain_mismatch": "Brand named but link does not match",
    "money_n": "Money terms (count)",
    "money_hit": "Mentions money transfer",
    "pressure_n": "Urgency terms (count)",
    "pressure_hit": "Uses urgency / pressure",
    "credentials_n": "Credential terms (count)",
    "credentials_hit": "Asks for credentials",
    "brands_n": "Brand mentions (count)",
    "brands_hit": "Mentions a known brand",
    "optout_n": "Opt-out phrases (count)",
    "optout_hit": "Carries legal opt-out language",
    "opener_n": "Relational openers (count)",
    "opener_hit": "Uses a relational opener",
    "length": "Message length",
    "n_words": "Word count",
    "capital_ratio": "Capital-letter ratio",
    "digit_ratio": "Digit ratio",
    "punct_ratio": "Punctuation ratio",
    "n_exclaim": "Exclamation marks",
    "n_allcaps_words": "ALL-CAPS words",
    "has_leetspeak": "Character substitution (M3dicare)",
    "avg_word_len": "Average word length",
}


# The features are a property of the dataset, but the parquet is a training input,
# so it is written next to the models that read it.
DATASET_CSV = os.path.join(HERE, "combined_sms_dataset.csv")
PARQUET = os.path.join(HERE, os.pardir, "scam-classification", "features.parquet")


def build(src=DATASET_CSV, out=PARQUET):
    import pandas as pd

    df = pd.read_csv(src, keep_default_na=False, dtype={"text": str})
    feats = pd.DataFrame([extract_features(t) for t in df["text"]], columns=FEATURE_NAMES)
    feats.insert(0, "id", df["id"].values)
    path = os.path.abspath(out)
    feats.to_parquet(path, index=False)
    return df, feats, path


if __name__ == "__main__":
    df, feats, path = build()
    print(f"{len(feats):,} rows x {len(FEATURE_NAMES)} features -> {path}\n")

    # Sanity view: mean of each feature per label. A feature that does not separate
    # smishing from ham here is unlikely to earn its place in the model.
    view = feats.drop(columns=["id"]).groupby(df["label"].values).mean().T
    view = view[[c for c in ("ham", "spam", "smishing") if c in view.columns]]
    print(view.round(3).to_string())

File: core/features.ts
/**
 * The 29 engineered features, ported from dataset/features.py.
 *
 * This file is the app's single largest correctness risk. Nothing here throws when it
 * is wrong -- a feature that drifts by one just moves the probability, and the app
 * keeps answering confidently. serving/app_assets/golden_vectors.json exists to catch
 * exactly that; run `node --experimental-strip-types test/golden.ts` after any edit.
 *
 * Places Python and JavaScript disagree, all of which bit during the port:
 *   - `len(s)` counts code points; `s.length` counts UTF-16 units, so one emoji is 2.
 *     Everything below iterates [...s] and uses codePoints().
 *   - Python's `\w` covers Unicode letters and digits; JavaScript's is ASCII-only.
 *     Word matching uses \p{L}\p{N}_ with the /u flag instead.
 *   - `str.isupper()` is true only when there is at least one cased character, so
 *     "123" is not all-caps. Uppercase comparison alone would say it is.
 *
 * Word lists are NOT duplicated here -- they load from feature_config.json, which
 * export_onnx.py writes from dataset/features.yaml.
 */

export type FeatureConfig = {
  known_tlds: string[];
  suspicious_tlds: string[];
  shorteners: string[];
  mismatch_brands: string[];
  families: Record<string, string[]>;
  keywords: string[];
};

const URL_RE =
  /(https?:\/\/|www\.|\b[a-z0-9-]+\.(com|net|org|info|xyz|ru|cn|co|io|me|us|biz|top|link|click|vip|shop|site|online|icu|cc|tk|ml|ga|gq|buzz)\b)/i;
const URL_FULL_RE = /(https?:\/\/|www\.)?([a-z0-9][a-z0-9.-]*\.[a-z]{2,})(\/\S*)?/gi;
const IP_DOMAIN_RE = /\b(?:\d{1,3}\.){3}\d{1,3}\b/;
const LEET_RE = /\b[a-z]+[0-9]+[a-z]+\b/i;
const WORD_RE = /[\p{L}\p{N}_]+/gu;

const IS_LETTER = /\p{L}/u;      // Python str.isalpha
const IS_UPPER = /\p{Lu}/u;      // a cased character in upper case
const IS_DIGIT = /\p{Nd}/u;      // Python str.isdigit
const IS_ALNUM = /[\p{L}\p{N}]/u;
const IS_SPACE = /\s/u;

const codePoints = (s: string): string[] => Array.from(s);

/** Non-overlapping occurrences, matching Python's str.count. */
function countOccurrences(haystack: string, needle: string): number {
  if (!needle) return 0;
  return haystack.split(needle).length - 1;
}

/** Python str.isupper: all cased characters upper, and at least one cased. */
function isAllCaps(w: string): boolean {
  return w === w.toUpperCase() && IS_UPPER.test(w);
}

function registrable(host: string): string {
  const parts = host.split('.');
  return parts.length >= 2 ? parts.slice(-2).join('.') : host;
}

/** Hosts that look like real domains, mirroring features.py:_domains. */
function extractDomains(text: string, knownTlds: Set<string>): string[] {
  const out: string[] = [];
  URL_FULL_RE.lastIndex = 0;
  let m: RegExpExecArray | null;
  while ((m = URL_FULL_RE.exec(text)) !== null) {
    if (m[0] === '') {
      URL_FULL_RE.lastIndex++;   // zero-width match: the prefix group is optional
      continue;
    }
    const prefix = m[1];
    const host = (m[2] || '').toLowerCase().replace(/\.+$/, '');
    if (!host.includes('.')) continue;
    if (/^\d+$/.test(host.replace(/\./g, ''))) continue;      // "3.50" is not a host
    const tld = host.split('.').pop() as string;
    if (!prefix && !knownTlds.has(tld)) continue;             // "tomorrow.We" is not
    out.push(host);
  }
  return out;
}

export function extractFeatures(text: string, cfg: FeatureConfig): Record<string, number> {
  const t = text || '';
  const low = t.toLowerCase();
  const flat = low.replace(/[^a-z0-9]/g, '');
  const chars = codePoints(t);
  const n = chars.length || 1;
  const words = t.match(WORD_RE) ?? [];

  const knownTlds = new Set(cfg.known_tlds);
  const suspiciousTlds = new Set(cfg.suspicious_tlds);
  const shorteners = new Set(cfg.shorteners);
  const domains = extractDomains(t, knownTlds);
  const tlds = new Set(domains.map((d) => d.split('.').pop() as string));

  const f: Record<string, number> = {};

  // --- links
  f.has_url = URL_RE.test(t) ? 1 : 0;
  f.n_urls = domains.length;
  f.has_shortener = domains.some((d) => shorteners.has(registrable(d))) ? 1 : 0;
  f.suspicious_tld = [...tlds].some((x) => suspiciousTlds.has(x)) ? 1 : 0;
  f.ip_as_domain = IP_DOMAIN_RE.test(t) && domains.length > 0 ? 1 : 0;

  const domainChars = domains.reduce((a, d) => a + codePoints(d).length, 0);
  const domainDigits = domains.reduce(
    (a, d) => a + codePoints(d).filter((c) => IS_DIGIT.test(c)).length, 0);
  f.url_digit_ratio = domainChars ? domainDigits / domainChars : 0.0;
  f.domain_len_max = domains.length
    ? Math.max(...domains.map((d) => codePoints(d).length)) : 0;

  const named = cfg.mismatch_brands.filter((b) => flat.includes(b));
  f.brand_domain_mismatch =
    named.length > 0 &&
    domains.length > 0 &&
    !named.some((b) => domains.some((d) => d.replace(/\./g, '').replace(/-/g, '').includes(b)))
      ? 1 : 0;

  // --- keyword families
  for (const [fam, terms] of Object.entries(cfg.families)) {
    const hits = terms.reduce((a, term) => a + countOccurrences(low, term), 0);
    f[`${fam}_n`] = hits;
    f[`${fam}_hit`] = hits > 0 ? 1 : 0;
  }

  // --- text shape
  const letters = chars.filter((c) => IS_LETTER.test(c)).length;
  f.length = chars.length;
  f.n_words = words.length;
  f.capital_ratio = chars.filter((c) => IS_UPPER.test(c)).length / (letters || 1);
  f.digit_ratio = chars.filter((c) => IS_DIGIT.test(c)).length / n;
  f.punct_ratio =
    chars.filter((c) => !IS_ALNUM.test(c) && !IS_SPACE.test(c)).length / n;
  f.n_exclaim = countOccurrences(t, '!');
  f.n_allcaps_words = words.filter(
    (w) => codePoints(w).length > 2 && isAllCaps(w)).length;
  f.has_leetspeak = LEET_RE.test(t) ? 1 : 0;
  f.avg_word_len = words.length
    ? words.reduce((a, w) => a + codePoints(w).length, 0) / words.length : 0.0;

  return f;
}

/** The 38 keyword flags, ported from dataset/keywords.py. */
export function keywordFlags(text: string, cfg: FeatureConfig): Record<string, number> {
  const low = (text || '').toLowerCase();
  const out: Record<string, number> = {};
  for (const kw of cfg.keywords) out[keywordColumn(kw)] = low.includes(kw) ? 1 : 0;
  return out;
}

/** dataset/keywords.py:column -- kw_ + the term slugged to [a-z0-9_]. */
export function keywordColumn(kw: string): string {
  return 'kw_' + kw.toLowerCase()
    .replace(/[^a-z0-9]+/g, '_')
    .replace(/_+/g, '_')
    .replace(/^_+|_+$/g, '');
}

/**
 * The model's input row, ordered by the manifest's feature_names.
 *
 * Assembled by name, never by concatenation -- the same rule serving/predict.py
 * follows. A column-order mismatch between training and the app does not raise; it
 * silently changes every answer.
 */
export function buildRow(
  featureNames: string[],
  embedding: Float32Array | number[] | null,
  feats: Record<string, number>,
  kws: Record<string, number>,
): Float32Array {
  const row = new Float32Array(featureNames.length);
  featureNames.forEach((name, i) => {
    if (name.startsWith('emb')) {
      if (!embedding) throw new Error(`${name} needs an embedding but none was given`);
      row[i] = embedding[Number(name.slice(3))];
    } else if (name in feats) {
      row[i] = feats[name];
    } else if (name in kws) {
      row[i] = kws[name];
    } else {
      throw new Error(`no value for feature "${name}"`);
    }
  });
  return row;
}
